src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -3468,3 +3445,355 @@ void c_typecheck_baset::make_constant_index(exprt &expr)
     throw 0;
   }
 }
+
+optionalt<symbol_exprt> c_typecheck_baset::typecheck_gcc_polymorphic_builtin(
+  const irep_idt &identifier,
+  const exprt::operandst &arguments,
+  const source_locationt &source_location)
+{
+  // the arguments need not be type checked just yet, thus do not make
+  // assumptions that would require type checking
+
+  if(
+    identifier == "__sync_fetch_and_add" ||
+    identifier == "__sync_fetch_and_sub" ||
+    identifier == "__sync_fetch_and_or" ||
+    identifier == "__sync_fetch_and_and" ||
+    identifier == "__sync_fetch_and_xor" ||
+    identifier == "__sync_fetch_and_nand" ||
+    identifier == "__sync_add_and_fetch" ||
+    identifier == "__sync_sub_and_fetch" ||
+    identifier == "__sync_or_and_fetch" ||
+    identifier == "__sync_and_and_fetch" ||
+    identifier == "__sync_xor_and_fetch" ||
+    identifier == "__sync_nand_and_fetch" ||
+    identifier == "__sync_bool_compare_and_swap" ||
+    identifier == "__sync_val_compare_and_swap" ||
+    identifier == "__sync_lock_test_and_set" ||
+    identifier == "__sync_lock_release")
+  {
+    // These are polymorphic, see
+    // http://gcc.gnu.org/onlinedocs/gcc-4.1.1/gcc/Atomic-Builtins.html
+
+    // adjust return type of function to match pointer subtype
+    if(arguments.empty())
+    {
+      error().source_location = source_location;
```

Comment:
consider using proper exception

Reply:
Yes, but not in this PR I'd say. We need to fix this all over the C/C++ front-end.